llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp

```diff
@@ -977,6 +977,22 @@ void SelectionDAGLegalize::LegalizeOp(SDNode *Node) {
   TargetLowering::LegalizeAction Action = TargetLowering::Legal;
   bool SimpleFinishLegalizing = true;
   switch (Node->getOpcode()) {
+  // FIXME: If the node represents a poison value, replace it with an undef
+  // value.
+  //  A poison value results from an erroneous operation but does not cause
+  //  immediate undefined behavior, allowing speculative execution.
+  //  Since most operations propagate poison, it is valid to replace poison
+  //  with an undef value, which can take any legal value of the same type.
+  //  This ensures that downstream computations do not rely on poison semantics.
+  //  Poison is more restrictive than undef. Since we replace poison with undef
+  //  here, the poison information will be lost after the code is executed. In
+  //  the futher, If we need to retain the poison information after the code is
```

Comment:
```suggestion
  //  the future, If we need to retain the poison information after the code is
```

Reply:
But despite the big comment, I don't really understand why this is necessary. Why isn't poison going through the usual legalization here (you already added various support for it?)

Reply:
"after the code is execute" - is this referring to the code in this function in the compioler or the code in the program being compiled?

Reply:
> Why isn't poison going through the usual legalization here (you already added various support for it?)

I suspect most of the added type legalization is creating UNDEF instead of POISON. I only checked `SplitRes_UNDEF` which calls getUNDEF.